### ISSUE TITLE
Does mitmproxy support bandwidth throttling?

### ISSUE THREAD

**Issue by PankajKaushik78** [mitmproxy/mitmproxy#5208]
Hey, is there any way we can do bandwidth throttling to simulate the slow internet connections (2G, 3G, 4G)?

**Comment by mhils** [comment]
There's no functionality for that at the moment, but it's definitely something we could add. :)

**Comment by himanshusharma2973** [comment]
@mhils by when can we expect this feature ready?

**Comment by himanshusharma2973** [comment]
@mhils Or how can we achieve this with the current setup

**Comment by mhils** [comment]
I have not plans to work on this myself anytime soon.

**Comment by istepanko** [comment]
This feature is highly anticipated(Charles has it) especially when there is no privileged docker execution available(in large orgs) there is no other way to throttle traffic for testing on k8s(Cilium).

**Comment by matthewmturner** [comment]
@mhils do you have any advice on where someone new to the project would start if they wanted to implement throttling?

**Comment by mhils** [comment]
Our core I/O happens in https://github.com/mitmproxy/mitmproxy/blob/main/mitmproxy/proxy/server.py. It probably makes sense to add it there. Alternatively we could consider a custom layer, but it's more tricky to add back pressure there.

**Comment by matthewmturner** [comment]
Thanks @mhils!

Would a reasonable approach be buffering `data` in the below section and handing it off to `self.server_event` based on whatever bandwidth is selected by user?

https://github.com/mitmproxy/mitmproxy/blob/849a3c33cb10670eeab7f77a797c544f38e09d6b/mitmproxy/proxy/server.py#L280-L297

**Comment by mhils** [comment]
(I haven't spend a lot of time thinking about this, but here are my two cents)

This looks like it's going in the right direction. Generally speaking I think we need to slow down the reading operations instead of buffering data here. The problem is that we otherwise don't create any backpressure. Here's an example of what I mean:

1. Client sends an large HTTP POST at time *t*. Mitmproxy is instructed to stream the body.
2. mitmproxy (on the same machine) reads it all off the socket by *t+2s*, buffers it internally, and slowly releases it to `self.server_event` over the next two minutes.
3. At *t+62*, the client believes the server is processing for a minute now and closes the connection/starts retries. In reality, the target server has only received half of the contents so far.

To avoid this situation, we need to create backpressure, i.e. keep the client <-> proxy TCP pipe full. The way that's currently done is by calling `drain_writers()` after every read in the snippet you posted above. Adding more pausing there seems like the natural way to go.

**Comment by matthewmturner** [comment]
thank you for the explanation, @mhils.

I will look into this.

**Comment by AnilPatidar** [comment]
Here chatGPT found a solutions to this problem

Here is an example Python script that you can use with mitmdump to throttle network traffic:

python
Copy code
import time
import mitmproxy.http

class Throttle:
    def __init__(self, bandwidth):
        self.bandwidth = bandwidth
        self.bytes_sent = 0
        self.start_time = None

    def request(self, flow: mitmproxy.http.HTTPFlow):
        self.start_time = time.monotonic()

    def response(self, flow: mitmproxy.http.HTTPFlow):
        elapsed_time = time.monotonic() - self.start_time
        self.bytes_sent += len(flow.response.content)
        if elapsed_time > 0:
            bandwidth = self.bytes_sent / elapsed_time
            if bandwidth > self.bandwidth:
                time.sleep(self.bytes_sent / self.bandwidth - elapsed_time)
                self.bytes_sent = 0
                self.start_time = None

addons = [
    Throttle(100 * 1024)  # 100 KB/s
]
In this example, the Throttle class is defined with a constructor that takes a bandwidth parameter to specify the desired bandwidth limit in bytes per second. The request method is called when a request is received and the response method is called when a response is sent. The bandwidth usage is calculated and compared to the specified limit, and if the limit is exceeded, the script will sleep for the appropriate amount of time to throttle the network traffic.

To use this script with mitmdump, simply save the script to a file (e.g. throttle.py) and run mitmdump with the -s option to specify the script file:

Copy code
mitmdump -s throttle.py
Note that this script will only throttle the traffic for the mitmproxy instance running on the same machine. If you need to throttle network traffic for all devices on your network, you will need to set up a proxy server that routes all traffic through mitmproxy, and configure the devices to use the proxy server.

**Comment by mhils** [comment]
> Can that be done easily with `next_layer` and a layer with custom `_handle_event`?

Yes, pretty much. You need to implement `next_layer` in your addon (we do have an example for this [here](https://github.com/mitmproxy/mitmproxy/blob/10.1.3/examples/contrib/custom_next_layer.py) - check out the builtin [NextLayer addon](https://github.com/mitmproxy/mitmproxy/blob/10.1.3/mitmproxy/addons/next_layer.py) as well), and then assign a custom layer instance to `nextlayer.layer`. In your throttling layer, you probably want to buffer all incoming data and then gradually forward it to your sublayer. You can use Wakeup/RequestWakeup to get your layer called when it's time to shove in more bytes (see [here](https://github.com/mitmproxy/mitmproxy/blob/10.1.3/mitmproxy/proxy/layers/http/_http2.py#L506)). If you build something along those lines, I'd appreciate a PR for examples/contrib! 😃

**Comment by xueli181114** [comment]
Hi guys, I happen to find this issue. Is there any ongoing PR for the implementation?

**Comment by rahimrahman** [comment]
In the meantime, if you're on MacOS, you can throttle using pfctl and dnctl.  Probably something similar on Linux machine.

```
echo "dummynet in from www.example.com to any pipe 1
dummynet out from any to www.example.com" | sudo pfctl -f -

# pipe 1 is download, below is 3.3Mb/s with 1000ms latency
sudo dnctl pipe 1 config bw 3300Kbit/s delay 1000ms

# pipe 2 is upload, below is 3.3Mb/s with 1000ms latency
sudo dnctl pipe 2 config bw 3300Kbit/s delay 1000ms
sudo pfctl -E
```

```
# and to disable
sudo pfctl -d
```
